standards/practices/testing.md

```diff
@@ -0,0 +1,198 @@
+{
+    use RefreshDatabase;
+
+    /** @test */
+    public function admins_can_log_in()
+    {
+        // Arrange
+        // Create an admin.
+        $user = factory(User::class)->create(['password' => bcrypt('pw1234')]);
+
+        // Act
+        // Try to log in.
+        $response = $this->post('/login', [
+            'email'    => $user->email,
+            'password' => 'pw1234'
+        ]);
+
+        // Assert
+        // Ensure user logged in.
+        $response->assertRedirect(route('home'));
+        $this->assertAuthenticatedAs($user);
+    }
+
+    /** @test */
+    public function client_users_can_log_in()
+    {
+        // Arrange
+        // Create a client user.
+        $user = factory(ClientUser::class)->create(['password' => bcrypt('pw1234')]);
+
+        // Act
+        // Try to log in.
+        $response = $this->post('/login', [
+            'email'    => $user->email,
+            'password' => 'pw1234'
+        ]);
+
+        // Assert
+        // Ensure user logged in.
+        $response->assertRedirect(route('home'));
+        $this->assertAuthenticatedAs($user);
+    }
+}
+```
+
+*In most frameworks, it's not worth testing authentication, because the framework handles pretty much everything. However in this example, the authentication is customized enough that we need to avoid regressions.*
+
+These two tests improve on the previous example in several ways:
+
+1. We converted from a unit test to an integration test (called a feature test in Laravel). The feature depends on several systems more simply tested with a couple short integration tests than with several unit tests that can't reliably test the whole feature.
+2. Each test covers only one action.
+3. We formatted the test names so people can easily read it and understand what it tests.
+4. Each test is split into the standard arrange, act, and assert steps with helpful explanations.
+5. The tests cover complex functions likely affected by regressions.
+
+
+## Explanation
+Tests help us improve the quality of our software. However, tests carry maintenance costs, so **our tests should add value to our software**. In pursuit of this value:
+
+1. **Avoid overly-specific coverage goals.**  
+  E.g. 80% code coverage. Goals like this incentivize writing more tests than required to obtain confidence in the software.
+2. **Don't test what you don't own.**  
+  Testing languages, frameworks, and other third-party software indicates you don't trust the dependency you decided to rely on.
+3. **Avoid testing basic operations.**  
+  E.g. Testing an Eloquent model's relationships. Though you own the method on the model, in nearly every case, it merely consumes the framework.
+4. **Test the input and output, not the implementation.**  
+  Implemtation may change over time. Tests shouldn't fail after refactoring.
+
+Breaking these guidelines usually adds to maintenance costs without adding worthwhile value to our software.
+
+### Unit tests
```

Comment:
I recommend adding a paragraph about the testing pyramid of Unit, Feature, and Acceptance which is the typical way to categorize back-end tests. Then also add a bit about the pyramid of Unit, Snapshot, and e2e (end-to-end) tests that are typical of front-end and especially js apps. Edd Yerburgh in his book "Testing Vue.js Applications" talks about this on p194. Snapshot tests are a different category but  suggest noting the different pyramids too. Front-end and back-end tests are different enough to  benefit from guidance for the developer.

Reply:
I don't find pyramids practically useful. Are we going to tally the tests in each of our categories and compare them to a pyramid? That seems valueless to me. If we're not going to do something along those lines, what purpose does including pyramids serve?

I'm opposed to adding anything we don't have sufficient experience in to a standard. I'd say given the only snapshot tests we've written are on a single project, and we're still struggling to understand them and how they can provide value, we don't have sufficient experience.

That said, we don't need a standard in order to use them, and we have a process for amending standards. I suggest we continue using them and once we have a good understanding of how to create value with them, we can amend this standard.

Reply:
That sounds fine. We are inexperienced at this point. We can add more later. btw, I think the pyramid thing is helpful since it comes out of a lot of other people's experience in the larger world of testing.

Reply:
Yeah, I think the pyramid can be a helpful guideline when first getting started with testing, but beyond that it seems rather arbitrary to me. I don't really see a practical use for it, which is all I'm really interested in for standards.

Plus, even the pyramid is now debated within the dev community. A lot of people are pushing a "testing trophy" now. I don't really think either truly applies globally enough to form a standard around them.

Reply:
In general I think these standards will help us better if we try to set them as the base level we all should be following. We shouldn't add untested, or new to us methods to a standard. It should be more of a baseline that we all understand and operate from, as opposed to as much of an inspirational goal or a possible future state. 
I think it is a sign of a problem if people are getting _introduced_ to a concept through a standard as opposed to it reinforcing ideas we are already committed to.